File: benchmarks/livermore_canonical/k12_first_diff.f
C     Kernel 12 -- First difference (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body: DO k=1,n: X(k) = Y(k+1) - Y(k)
C
      PROGRAM K12
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION X(1000), Y(1001)
C
C     Initialize Y
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Y(K) = (BUZZ - FIZZ) * 0.1D0
   10 CONTINUE
C
      N = 1000
      DO 200 REP = 1, 35000000
          DO 90 K = 1, N
              X(K) = Y(K+1) - Y(K)
   90     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) X(1)
  900 FORMAT('x(1) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k12_first_diff.c
/* K12 -- First difference (canonical, 1-based indexing)
   Canonical body: DO k=1,n: X(k) = Y(k+1) - Y(k) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1000
#define NREPS 41000000

int main(void) {
    double x[1001], y[1002];

    signel(y+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        for (long k = 1; k <= N; k++) {
            x[k] = y[k+1] - y[k];
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("x[1] = %f\n", x[1]);
    return 0;
}
